dashboard/lib/pd/foorm/foorm_parser.rb

```diff
@@ -0,0 +1,119 @@
+# Parses Foorm forms into a useful format for looking up questions by form name, form version
+# and question name
+# Format is:
+# {
+#   <form-name>.<form-version: {
+#     <question_name>: {
+#       title: "sample title",
+#       type: "text/singleSelect/multiSelect/matrix/scale",
+#       # for singleSelect/multiSelect/scale
+#       choices: {
+#         choice_1_name: "choice 1 value",
+#         ...
+#       },
+#       # if has other choice
+#       other_text: "Other choice text",
+#       # for matrix
+#       rows: {
+#         row_1_name: "row 1 value",
+#         ...
+#       },
+#       columns: {
+##         column_1_name: "column 1 value",
+##         ...
+##       }
+#     }
+#   }
+# }
+module Pd::Foorm
+  class FoormParser
+    include Constants
+    extend Helper
+
+    # parse all forms in given list and return object in format above
+    def self.parse_forms(forms)
+      parsed_forms = {}
+      forms.each do |form|
+        parsed_form = {}
+        form_questions = JSON.parse(form.questions, symbolize_names: true)
+        form_questions[:pages].each do |page|
+          page[:elements].each do |question_data|
+            parsed_form = parse_element(question_data, parsed_form)
+          end
+        end
+        parsed_forms[get_form_key(form.name, form.version)] = parsed_form
+      end
+      parsed_forms
+    end
+
+    # parse a form element and add to existing parsed_form object
+    # Form element may be a panel which contains questions
+    def self.parse_element(question_data, parsed_form)
```

Comment:
the reason I wrote it recursively is a form can have infinitely nested panels, which was cleanest to go through recursively. I can try to translate to a loop but it will be messier

Reply:
updated to keep recursion but not alter input, instead return hash of results that gets merged into overall hash

Reply:
This looks great!  And nice idea @hacodeorg.